Publish portable build

RussKie opened the pull request:
<!-- Please read CONTRIBUTING.md before submitting a pull request -->




## Proposed changes

- Effectively revert a606467bb23721313b211b004b3456ce30572969, and publish a portable build

NB: the plugins are not included in the portable build right now



----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

RussKie commented:
A portable build now again - https://ci.appveyor.com/project/gitextensions/gitextensions/builds/30847034/artifacts

Attempting to start an app I get the following error:
![image](https://user-images.githubusercontent.com/4403806/74646643-59354e80-51ce-11ea-889a-8a82a2753014.png)

The error is not fatal, it complains about absent `Theme\current.colors` file, which gets created as part of the launch sequence.

@NikolayXHD any thoughts what needs to be done to improve the UX?

NikolayXHD commented:
> @NikolayXHD any thoughts what needs to be done to improve the UX?

#7562 gets rid of `current.colors` and the popups

NikolayXHD commented:
Found 2 issues

1. I broke dark theme for portable version. Never tested color settings migration with portable build and **!!!surprise!!!** I broke dark theme completely. At least the missing file popup is gone, as I promised.

2. `EasyHook32.dll`, `EasyHook64.dll` are missing in portable .zip
`MakePortableArchive.cmd` still points to correct files, so currently I have no explanation for this.

Will be looking at it closely the coming weekend.

RussKie commented:
Thank you for testing, more work to do.

On Wed, Feb 19, 2020, 12:01 PM NikolayHD <notifications@github.com> wrote:

> Found 2 issues
>
>    1.
>
>    I broke dark theme for portable version. Never tested color settings
>    migration with portable build and *!!!surprise!!!* I broke dark theme
>    completely. At least the missing file popup is gone, as I promised.
>    2.
>
>    EasyHook32.dll, EasyHook64.dll are missing in portable .zip
>    MakePortableArchive.cmd still points to correct files, so currently I
>    have no explanation for this.
>
> Will be looking at it closely the coming weekend.
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/gitextensions/gitextensions/pull/7785?email_source=notifications&email_token=ABBTEXTEWXIEG5UVPVRNN7TRDSAH5A5CNFSM4KWPGUO2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEMF6HHQ#issuecomment-587981726>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABBTEXREEUUYXRAP6I3JKGDRDSAH5ANCNFSM4KWPGUOQ>
> .
>

RussKie commented:
> 2\. I have no explanation for this.

I'm running a "publish" target against GitExtensions.csproj, that only copies files that are marked as "referenced" by the project. EasyHook dlls aren't technically referenced, and hence they don't get copied. That's my hypothesis.

RussKie commented:
I tweaked the publish target to include EasyHook dlls. I further tweaked it to include a build version in the name of the zip. And, while at it, I rebased on top of the latest master.
Please let me know if anything else I missed or is broken.

> The old portable builds had an additional directory inside

fixed



> 2\. `EasyHook32.dll`, `EasyHook64.dll` are missing in portable .zip

fixed

RussKie commented:
🎉 [https://ci.appveyor.com/api/buildjobs/gd3jsouxlbhsfxq1/artifacts/artifacts%2Fpublish%2FPortable-4.0.0.8598.zip](https://ci.appveyor.com/project/gitextensions/gitextensions/builds/30907552/artifacts)

pmiossec commented on GitExtensions/Project.Publish.targets:
```diff
@@ -0,0 +1,84 @@
+      <_PublishAppConfigPath>$([System.IO.Path]::Combine('$(PublishDir)', '$(_PublishAppConfig)'))</_PublishAppConfigPath>
+
+      <!-- Resolve the output file -->
+      <_PublishPortableFileName>Portable-$(CurrentBuildVersion.ToString()).zip</_PublishPortableFileName>
```
Is it possible to add 'GitExt' or 'GitExtensions' at the beginning of the zip file.
I think that will make it more identifiable in a full download folder ;)

And also **if it's really straight forward** to do, add the hash at the end by following the SemVer syntax (i.e  `GitExtensions-Portable-4.0.0.8598+f0d22742.zip`)

RussKie replied:
Yes, it is easy to add a prefix:
```xml
<_PublishPortableFileName>GitExtensions-Portable-$(CurrentBuildVersion.ToString()).zip</_PublishPortableFileName>
```

A hash may be a (little?) more involved, but doable.

RussKie replied:
Done. 
For non-release builds I propose we just have a commit hash instead of a build number, e.g. **GitExtensions-portable-f511a73a7.zip**. For a release we'll have a build number, e.g. **GitExtensions-portable-3.3.0.2342.zip**.
What do you think?

pmiossec replied:
I think that's good for now.

It will be perfect for 'master' builds and releases. No so much for PR (I thought about it afterwards...) 

I will try to look **later** at how to  retrieve the hash of the head of the pull request and use it as the hash here.

Because Appveyor build a commit which is a merge of the PR within 'master' so the hash will be of no use in this case 😕

pmiossec replied:
https://www.appveyor.com/docs/environment-variables/

* APPVEYOR_PULL_REQUEST_HEAD_COMMIT - Pull (Merge) Request source commit ID (SHA)

Or more likely :

* APPVEYOR_REPO_COMMIT - commit ID (SHA)

I think that we should add something like that to override the value on Appveyor (in pseudo msbuild on smartphone 😁) :

`<_PublishPortableFileNameSuffix Condition="'$(GitCommit)' != '' And '$(env:APPVEYOR_REPO_COMMIT)' !='' ">-$(env:APPVEYOR_REPO_COMMIT)</_PublishPortableFileNameSuffix>`

mstv replied:
> For non-release builds I propose we just have a commit hash instead of a build number, e.g. GitExtensions-portable-f511a73a7.zip. For a release we'll have a build number, e.g. GitExtensions-portable-3.3.0.2342.zip.

Build numbers have the advantage that they can be used for a kind of bisecting without building the project. I'd prefer to have it with build number and with hash appended.

> dirty

I agree only if AppVeyor builds a temporary merge with master.
But I don't like this for two reasons:
1. AppVeyor denies to build if there are conflicts with master. So you cannot finish the feature without being forced to merge the master branch.
2. The resulting build is a moving target depending on the master branch.
I'd prefer that AppVeyor builds exactly the _clean_ commit for PRs.

RussKie replied:
> Build numbers have the advantage that they can be used for a kind of bisecting without building the project. I'd prefer to have it with build number and with hash appended.

Apologies, I don't follow the logic here. How is a build number useful for bisecting, e.g. 4.0.0.1234? I can see a hash being useful though.



> I'd prefer that AppVeyor builds exactly the _clean_ commit for PRs.

The current function of "dirty" flag is to indicate whether a build is an official release or not. It is useful when users send us bug reports.
If we mark all builds as "clean" we won't know whether a specific build is official or not _without going to the GitHub release tab and checking there_.

> 1. AppVeyor denies to build if there are conflicts with master

If there is a conflict, it has to be resolved one way or another. I don't follow how it relates to the "dirty" flag discussion, sorry.

mstv replied:
> bisecting

Usually, I have many downloaded portable builds. They are well-ordered by build numbers.
Checking different builds for an issue with commits hashs only, would be a pain.

> dirty

My understanding of "dirty" is "There are uncommitted changes."
If I really forgot the build number of a release, I will see it in my folder with the portable builds (last one).

> If there is a conflict, it has to be resolved one way or another.

Yes, but we should allow to finish the implementation and being able to check it by AppVeyor before it is merged with master.

> I don't follow how it relates to the "dirty" flag discussion, sorry.

If AppVeyor builds a temporary merge with master, this _is_ a "dirty" build because it is no official commit.

RussKie replied:
> this _is_ a "dirty" build because it is no official commit.

You mis-understand, 'clean' builds only come from release branches. All other commits are marked dirty. This is how we had it since we introduced GitInfo into our builds.

A 'dirty' build means there are uncommitted changes in the working directory. We inject build numbers into AssemblyInfo and C++ header files, and that makes builds 'dirty'. For release builds (non-pr commits on 'release' branches) we ignore these changes making builds 'clean'.


> Usually, I have many downloaded portable builds. They are well-ordered by build numbers.
> Checking different builds for an issue with commits hashs only, would be a pain.

Make sense.

mstv replied:
I don't mis-understand this IMO. I just had not in mind, that a non-release build is modified in this way and hence is really a "dirty" build from git point of view. Though IMO this is just a technical issue how we inject build numbers. Whether this makes a commit build dirty or not, is a non-productive philosophical question. I can live with "dirty" non-release builds.

pmiossec replied:
> I agree only if AppVeyor builds a temporary merge with master.
>[...] 
>I'd prefer that AppVeyor builds exactly the clean commit for PRs.

I think that is GitHub that trigger Appveyor builds only if there are no conflicts.
And also GitHub trigger it 2 times for each PR. One is to build only the PR. The other is for the PRs merged with 'master' (to do real Continuous Integration).

That makes me think of something important. What is the build result that we consider to be the 'dev' version and that we should propose to download and test.

I think that it's better to use the 'merged' one.

Perhaps also we should do something to differentiate the artifacts because they will have the same name 🤔

RussKie replied:
> And also GitHub trigger it 2 times for each PR. One is to build only the PR. The other is for the PRs merged with 'master' (to do real Continuous Integration).

This only happens for commits on the main trunk. It doesn't happen for PRs coming from forks.

pmiossec replied:
Thanks for the clarification.

gerhardol commented:
Build is dirty

Still no dark theme...
No theme to select
I tried to copy my current (unmodified but not master) dark.colors, then I get an exception opening Settings

```
System.InvalidOperationException: Themes were not added to ComboBox
   at GitUI.CommandsDialogs.SettingsDialog.Pages.ColorsSettingsPage.set_SelectedThemeId(ThemeId value)
   at GitUI.CommandsDialogs.SettingsDialog.Pages.ColorsSettingsPage.SettingsToPage()
   at GitUI.CommandsDialogs.SettingsDialog.SettingsPageBase.LoadSettings()
   at GitUI.CommandsDialogs.FormSettings.LoadSettings()
   at GitUI.CommandsDialogs.FormSettings.FormSettings_Shown(Object sender, EventArgs e)
   at System.Windows.Forms.Form.OnShown(EventArgs e)
   at System.Windows.Forms.Form.CallShownEvent()
   at System.Windows.Forms.Control.InvokeMarshaledCallbackDo(ThreadMethodEntry tme)
   at System.Windows.Forms.Control.InvokeMarshaledCallbackHelper(Object obj)
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Windows.Forms.Control.InvokeMarshaledCallback(ThreadMethodEntry tme)
   at System.Windows.Forms.Control.InvokeMarshaledCallbacks()
```


## Environment
- Git Extensions 4.0.0.8598
- Build 8896d5910e3933caf13b8c5eb6db8ac972f52641 (Dirty)
- Git 2.24.0.windows.2
- Microsoft Windows NT 6.2.9200.0
- .NET Framework 4.8.4075.0
- DPI 96dpi (no scaling)

vbjay commented:
You can use something like this

https://github.com/gitextensions/gitextensions/blob/master/Setup/Get-Hashes.ps1

On Wed, Feb 19, 2020, 5:42 PM Igor Velikorossov <notifications@github.com>
wrote:

> *@RussKie* commented on this pull request.
> ------------------------------
>
> In GitExtensions/Project.Publish.targets
> <https://github.com/gitextensions/gitextensions/pull/7785#discussion_r381589352>
> :
>
> > +  <!--
> +    ============================================================
> +                       CreatePortable
> +
> +    Creates a portable archive.
> +    ============================================================
> +    -->
> +
> +  <Target Name="CreatePortable" AfterTargets="Publish" DependsOnTargets="_RetrieveCurrentBuildVersion;_PublishTranslations;_PublishExtraDependecies">
> +    <PropertyGroup>
> +      <!-- Resolve app.config, so we can set/unset "portable" flag -->
> +      <_PublishAppConfig>$([System.IO.Path]::GetFileName('$(TargetAppConfig)'))</_PublishAppConfig>
> +      <_PublishAppConfigPath>$([System.IO.Path]::Combine('$(PublishDir)', '$(_PublishAppConfig)'))</_PublishAppConfigPath>
> +
> +      <!-- Resolve the output file -->
> +      <_PublishPortableFileName>Portable-$(CurrentBuildVersion.ToString()).zip</_PublishPortableFileName>
>
> Yes, it is easy to add a prefix:
>
> <_PublishPortableFileName>GitExtensions-Portable-$(CurrentBuildVersion.ToString()).zip</_PublishPortableFileName>
>
> A hash may be a (little?) more involved, but doable.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/gitextensions/gitextensions/pull/7785?email_source=notifications&email_token=AA3WDMJSTRUSSKJXKFYXBL3RDWYUBA5CNFSM4KWPGUO2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCWF7GAI#discussion_r381589352>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AA3WDMMCH7QMWQSY3JDSBADRDWYUBANCNFSM4KWPGUOQ>
> .
>

RussKie commented:
I think @pmiossec was referring to a commit hash.
Which we have, its retrieval just needs to be plumb in. I think we should
also bake the hash in to assembly version (one its attributes)

On Thu, Feb 20, 2020, 10:33 AM Jay Asbury <notifications@github.com> wrote:

> You can use something like this
>
>
> https://github.com/gitextensions/gitextensions/blob/master/Setup/Get-Hashes.ps1
>
> On Wed, Feb 19, 2020, 5:42 PM Igor Velikorossov <notifications@github.com>
> wrote:
>
> > *@RussKie* commented on this pull request.
> > ------------------------------
> >
> > In GitExtensions/Project.Publish.targets
> > <
> https://github.com/gitextensions/gitextensions/pull/7785#discussion_r381589352
> >
> > :
> >
> > > + <!--
> > + ============================================================
> > + CreatePortable
> > +
> > + Creates a portable archive.
> > + ============================================================
> > + -->
> > +
> > + <Target Name="CreatePortable" AfterTargets="Publish"
> DependsOnTargets="_RetrieveCurrentBuildVersion;_PublishTranslations;_PublishExtraDependecies">
> > + <PropertyGroup>
> > + <!-- Resolve app.config, so we can set/unset "portable" flag -->
> > +
> <_PublishAppConfig>$([System.IO.Path]::GetFileName('$(TargetAppConfig)'))</_PublishAppConfig>
> > + <_PublishAppConfigPath>$([System.IO.Path]::Combine('$(PublishDir)',
> '$(_PublishAppConfig)'))</_PublishAppConfigPath>
> > +
> > + <!-- Resolve the output file -->
> > +
> <_PublishPortableFileName>Portable-$(CurrentBuildVersion.ToString()).zip</_PublishPortableFileName>
> >
> > Yes, it is easy to add a prefix:
> >
> >
> <_PublishPortableFileName>GitExtensions-Portable-$(CurrentBuildVersion.ToString()).zip</_PublishPortableFileName>
> >
> > A hash may be a (little?) more involved, but doable.
> >
> > —
> > You are receiving this because you are subscribed to this thread.
> > Reply to this email directly, view it on GitHub
> > <
> https://github.com/gitextensions/gitextensions/pull/7785?email_source=notifications&email_token=AA3WDMJSTRUSSKJXKFYXBL3RDWYUBA5CNFSM4KWPGUO2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCWF7GAI#discussion_r381589352
> >,
> > or unsubscribe
> > <
> https://github.com/notifications/unsubscribe-auth/AA3WDMMCH7QMWQSY3JDSBADRDWYUBANCNFSM4KWPGUOQ
> >
> > .
> >
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/gitextensions/gitextensions/pull/7785?email_source=notifications&email_token=ABBTEXWBXZDQMMR5OWXTCJLRDW6TRA5CNFSM4KWPGUO2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEMKD6QQ#issuecomment-588529474>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABBTEXW4CCQXWTTJHO2OFE3RDW6TRANCNFSM4KWPGUOQ>
> .
>

RussKie commented:
> Build is dirty

Correct, we're not building a release (if that's what you mean)

> Still no dark theme...
> No theme to select

@NikolayXHD would you be able to check this? Likely I have missed something... 😕

gerhardol commented:
>> Build is dirty

> Correct, we're not building a release (if that's what you mean)

No server build should be git dirty

NikolayXHD commented:
> @NikolayXHD would you be able to check this? Likely I have missed something... 😕

Yes, I will. It's a mistake somewhere in my last PR. Will figure it out more or less by the weekend.

gerhardol commented:
Works for me
I would have preferred to be able to edit and save themes also in portable, even if it means that you can modify the built in themes. Certainly not a blocker.

NikolayXHD commented:
> I would have preferred to be able to edit and save themes also in portable, even if it means that you can modify the built in themes. Certainly not a blocker.

Theme editing is gone at the moment, it's not portable version specific. Had to sacrifice it to keep code quality decent at the moment of release.

Theme files are just plain text with hex color codes, so in principle it can be edited in text editor with coupled with some online palette tool, such as color.adobe.com

RussKie commented on GitUI/Theming/ThemeRepository.cs:
```diff
@@ -71,9 +86,19 @@ private IEnumerable<ThemeId> GetUserCustomizedThemeIds()
+
+            if (UserThemesDirectory == null)
+            {
+                throw new NotSupportedException("There is no directory for custom user themes in portable mode");
```
```suggestion
                throw new InvalidOperationException("There is no directory for custom user themes in portable mode");
```
https://docs.microsoft.com/dotnet/api/system.notsupportedexception#related-exception-types

RussKie commented on GitUI/Theming/ThemeRepository.cs:
```diff
@@ -71,9 +86,19 @@ private IEnumerable<ThemeId> GetUserCustomizedThemeIds()
-            Path.Combine(id.IsBuiltin
-                ? AppDirectory
-                : DataDirectory, id.Name + Extension);
+        private string GetPath(ThemeId id)
```
Is it technically possible to have a theme in a portable build that isn't a 'built-in'?

gerhardol replied:
Yes, you can still copy an existing theme and adapt that.
Name is set from the filename.

RussKie replied:
In that case, won't we be falling through to `UserThemesDirectory == null` code path?

gerhardol replied:
You would copy the theme to the app data directory, user data directory does not exist (really the same as app data).

RussKie commented:
> I would have preferred to be able to edit and save themes also in portable

The editor added a lot of code that likely won't be used by the majority of our users. With that it brought a lot of maintenance overhead for us.
We can probably release it as a plugin.

RussKie commented:
> No server build should be git dirty

Respectfully disagree. We only made clean builds for release.
[Here's a PR build](https://ci.appveyor.com/project/gitextensions/gitextensions/builds/29905385/artifacts) from few months ago:
```
- Git Extensions 4.0.0.7889
- Build 364baef68efb2ed8f67320d16544056e4dcc6c0f (Dirty)
- Git 2.16.1.windows.4 (recommended: 2.23.0 or later)
- Microsoft Windows NT 10.0.18363.0
- .NET Framework 4.8.4075.0
- DPI 144dpi (150% scaling)
```

I am yet to look at the "merge" or "release" pipeline. Let's deal with the "pr" pipeline first.

gerhardol commented:
>> No server build should be git dirty

> Respectfully disagree. We only made clean builds for release.
I am yet to look at the "release" pipeline. Let's deal with the "pr" pipeline first.

Future problem, agree not in this PR

RussKie commented:
@msftbot merge in 24 hours
